flink-tests/src/test/java/org/apache/flink/api/datastream/DataStreamBatchExecutionITCase.java

```diff
@@ -0,0 +1,159 @@
+
+/**
+ * Integration test for {@link org.apache.flink.api.common.RuntimeExecutionMode#BATCH} execution on
+ * the DataStream API.
+ *
+ * <p>We use a {@link MiniClusterWithClientResource} with a single TaskManager with 1 slot to verify
+ * that programs in BATCH execution mode can be executed in stages.
+ */
+public class DataStreamBatchExecutionITCase {
+	private static final int DEFAULT_PARALLELISM = 1;
+
+	@ClassRule
+	public static MiniClusterWithClientResource miniClusterResource = new MiniClusterWithClientResource(
+			new MiniClusterResourceConfiguration.Builder()
+					.setNumberTaskManagers(1)
+					.setNumberSlotsPerTaskManager(DEFAULT_PARALLELISM)
+					.build());
+
+	/**
+	 * We induce a failure in the last mapper. In BATCH execution mode the part of the pipeline
+	 * before the key-by should not be re-executed. Only the part after that will restart. We check
+	 * that by suffixing the attempt number to records and asserting the correct number.
+	 */
+	@Test
+	public void batchFailoverWithKeyByBarrier() throws Exception {
+
+		final StreamExecutionEnvironment env = getExecutionEnvironment();
+
+		DataStreamSource<String> source = env.fromElements("ciao", "ciao");
+
+		SingleOutputStreamOperator<String> mapped = source
+				.map(new SuffixAttemptId("a"))
+				.map(new SuffixAttemptId("b"))
+				.keyBy(in -> in)
+				.map(new SuffixAttemptId("c"))
+				.map(new OnceFailingMapper("d"));
+
+		List<String> result = mapped.executeAndCollect(2);
+
+		// only the operators after the key-by "barrier" are restarted and will have the "attempt 1"
+		// suffix
+		assertThat(result, contains("ciao-a0-b0-c1-d1", "ciao-a0-b0-c1-d1"));
+	}
+
+	@Ignore
+	@Test
+	public void stagedExecutionWithDifferentSlotGroups() throws Exception {
+
+		final StreamExecutionEnvironment env = getExecutionEnvironment();
+
+		DataStreamSource<String> source = env.fromElements("ciao", "ciao");
+
+		SingleOutputStreamOperator<String> mapped = source
+				.map(new SuffixAttemptId("a"))
+				.map(new SuffixAttemptId("b"))
+				.map(new SuffixAttemptId("c"))
+				.slotSharingGroup("foo")
+				.map(new SuffixAttemptId("d"))
+				.slotSharingGroup("bar");
+
+		List<String> result = mapped.executeAndCollect(2);
+
+		// only the operators after the key-by "barrier" are restarted and will have the "attempt 1"
+		// suffix
+		assertThat(result, contains("ciao-a0-b0-c0-d0", "ciao-a0-b0-c0-d0"));
+	}
+
+	private StreamExecutionEnvironment getExecutionEnvironment() {
+		final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+
+		Configuration config = new Configuration();
+		config.set(ExecutionOptions.RUNTIME_MODE, RuntimeExecutionMode.BATCH);
+		env.configure(config, DataStreamBatchExecutionITCase.class.getClassLoader());
+
+		env.setParallelism(1);
+		env.setRestartStrategy(RestartStrategies.fixedDelayRestart(10, Time.milliseconds(1)));
+
+		// trick the collecting sink into working even in the face of failures 🙏
```

Comment:
huh, emojis in comments, that's new ;) didn't know you can do that, nice

Reply:
yeah 😅